code/nrf-connect/samples/ble/src/encoding.c

```diff
@@ -0,0 +1,135 @@
+  out[5] = sensors->batt.millivolts & 0xff;
+  int16_t temp_centicelsius = 100 * sensors->shtc3.temp_c;
+  out[6] = temp_centicelsius >> 8;
+  out[7] = temp_centicelsius & 0xff;
+  uint16_t humi = sensors->shtc3.rel_humi * UINT16_MAX;
+  out[8] = humi >> 8;
+  out[9] = humi & 0xff;
+  uint16_t soil_moisture = sensors->soil.percentage * UINT16_MAX;
+  out[10] = soil_moisture >> 8;
+  out[11] = soil_moisture & 0xff;
+  // MAC address in big-endian.
+  memcpy(out + 12, bt_addr->a.val, BT_ADDR_SIZE);
+  out[18] = sensors->photo.brightness >> 8;
+  out[19] = sensors->photo.brightness & 0xff;
+
+// https://bthome.io/v1/
+#elif CONFIG_PRST_BLE_ENCODING_BTHOME_V1
+
+  out[0] = 0x1c;
+  out[1] = 0x18;
+
+  // 1. Soil moisture.
+  // uint16_t.
+  out[2] = (0b000 << 5) | 3;
+  // Type of measurement - Moisture.
+  out[3] = 0x14;
+  // Value. Factor of 0.01, so we need to multiply our the value in 100% by
+  // 1/0.01 = 100.
+  uint16_t soil_val = 10000 * sensors->soil.percentage;
+  out[4] = soil_val & 0xff;
+  out[5] = soil_val >> 8;
+
+  // 2. Temp.
+  // int16_t.
+  out[6] = (0b001 << 5) | 3;
+  // Type of measurement - temperature.
+  out[7] = 0x02;
+  // Value. Factor 0.01.
+  int16_t temp_val = 100 * sensors->shtc3.temp_c;
+  out[8] = temp_val & 0xff;
+  out[9] = temp_val >> 8;
+
+  // 3. Humidity
+  // uint16_t.
+  out[10] = (0b000 << 5) | 3;
+  // Type - humidity.
+  out[11] = 0x03;
+  // Value. Factor 0.01, over 100%.
+  uint16_t humi_val = 10000 * sensors->shtc3.rel_humi;
+  out[12] = humi_val & 0xff;
+  out[13] = humi_val >> 8;
+
+  // 4. Battery voltage.
+  // uint16_t.
+  out[14] = (0b000 << 5) | 3;
+  // Type - voltage.
+  out[15] = 0x0c;
+  // Value. Factor of 0.001.
+  uint16_t batt_val = sensors->batt.millivolts;
+  out[16] = batt_val & 0xff;
+  out[17] = batt_val >> 8;
+
+// https://bthome.io/format/
+#elif CONFIG_PRST_BLE_ENCODING_BTHOME_V2
+  // 0xfcd2 - bthome.io service UUID.
+  out[0] = 0xd2;
+  out[1] = 0xfc;
+
+  // Service header - no encryption, bt home v2.
+  out[2] = 0x40;
+
+  // Soil moisture.
+  out[3] = 0x14;
+  // Factor of 0.01, so we need to multiply our the value in 100% by 1/0.01 = 100.
+  uint16_t soil_val = 10000 * sensors->soil.percentage;
+  out[4] = soil_val & 0xff;
+  out[5] = soil_val >> 8;
+
+  // Temperature.
+  out[6] = 0x02;
+  int16_t temp_val = 100 * sensors->shtc3.temp_c;
+  out[7] = temp_val & 0xff;
+  out[8] = temp_val >> 8;
+
+  // Humidity.
+  out[9] = 0x03;
+  // Value. Factor 0.01, over 100%.
+  uint16_t humi_val = 10000 * sensors->shtc3.rel_humi;
+  out[10] = humi_val & 0xff;
+  out[11] = humi_val >> 8;
+
+  // Battery voltage.
+  out[12] = 0x0c;
+  // Value. Factor of 0.001.
+  uint16_t batt_val = sensors->batt.millivolts;
+  out[13] = batt_val & 0xff;
+  out[14] = batt_val >> 8;
+
+  // Illuminance.
+  out[15] = 0x05;
+  out[16] = sensors->photo.brightness >> 16;
+  out[17] = (sensors->photo.brightness >> 8) & 0xff;
+  out[18] = sensors->photo.brightness & 0xff;
```

Comment:
With `CONFIG_PRST_BLE_ENCODING_BTHOME_V2=y` HA logs this warning:

```
[bthome_ble.parser] BTHome device is not sending object ids in numerical order (from low to high object id). This can cause issues with your BTHome receiver, payload: 14ce0002b5080350180c660c05000000
```

Simple reordering of the measurement blocks gets rid of it. Measurements are shown either way:

![Screenshot from 2022-12-02 23-19-52](https://user-images.githubusercontent.com/98131/205396122-2cde19b9-b0a6-4a76-9400-8399266aeecc.png)

Reply:
I hadn't noticed issues with illuminance readings, but they seem fine now. Thanks!